src/Frontend/Core/Engine/Breadcrumb.php

```diff
@@ -40,48 +40,64 @@ public function __construct(KernelInterface $kernel)
         // add homepage as first item (with correct element)
         $this->addElement($homeInfo['navigation_title'], Navigation::getURL(1));
 
-        // get other pages
-        $pages = $this->URL->getPages();
+        $this->addBreadcrumbsForPages($this->URL->getPages());
+    }
+
+    /**
+     * @param array $pages
+     */
+    private function addBreadcrumbsForPages(array $pages)
+    {
+        $breadcrumbs = $this->getBreadcrumbsForPages($pages);
 
-        // init vars
-        $items = array();
+        // reverse so everything is in place
+        krsort($breadcrumbs);
+
+        // add the breadcrumbs
+        array_map(
+            function (array $breadcrumb) {
+                $this->addElement($breadcrumb['title'], $breadcrumb['url']);
+            },
+            $breadcrumbs
+        );
+    }
+
+    /**
+     * @param array $pages
+     *
+     * @return array
+     */
+    private function getBreadcrumbsForPages(array $pages): array
+    {
+        $breadcrumbs = array();
         $errorURL = Navigation::getURL(404);
 
         // loop pages
         while (!empty($pages)) {
-            // init vars
             $url = implode('/', $pages);
             $menuId = Navigation::getPageId($url);
             $pageInfo = Navigation::getPageInfo($menuId);
 
-            // do we know something about the page
-            if ($pageInfo !== false && isset($pageInfo['navigation_title'])) {
-                // only add pages that aren't direct actions
-                if ($pageInfo['tree_type'] != 'direct_action') {
-                    // get URL
-                    $pageURL = Navigation::getURL($menuId);
-
-                    // if this is the error-page, so we won't show an URL.
-                    if ($pageURL == $errorURL) {
-                        $pageURL = null;
-                    }
-
-                    // add to the items
-                    $items[] = array('title' => $pageInfo['navigation_title'], 'url' => $pageURL);
-                }
+            // if we don't have info for the page, a navigation title or if it is a direct action => skip the page
+            if ($pageInfo === false
+                || !isset($pageInfo['navigation_title'])
+                || $pageInfo['tree_type'] === 'direct_action') {
+                array_pop($pages);
+
+                continue;
             }
 
-            // remove element
-            array_pop($pages);
-        }
+            $pageURL = Navigation::getURL($menuId);
 
-        // reverse so everything is in place
-        krsort($items);
+            // if this is the error-page, so we won't show an URL.
+            if ($pageURL === $errorURL) {
+                $pageURL = null;
+            }
 
-        // loop and add elements
-        foreach ($items as $row) {
-            $this->addElement($row['title'], $row['url']);
+            $breadcrumbs[] = ['title' => $pageInfo['navigation_title'], 'url' => $pageURL];
```

Comment:
If you use array_push() to add one element to the array it's better to use $array[] = because in that way there is no overhead of calling a function.